Add helpers to install & remove katello-ca, update docker container tests

abalakh opened the pull request:
This PR:
* moves `install_katello_ca` from `robottelo/vm.py` to `robottelo/helpers.py`
* makes `install_katello_ca` accept optional argument `hostname` to be able to install certificate on any host
* adds `remove_katello_ca` helper
* updates cli content-host tests to use new helper
* adds new api docker container test, which requires katello_ca to be installed
* updates api/cli/ui docker tests to install & remove katello_ca before/after tests execution accordingly
* skips ui docker container tests, which are failing due to opened BZ

Closes #3044

abalakh commented:
Ran 1 test for API/CLI/UI on machine with downgraded docker & commented `@skip_if_bug_open('bugzilla', 1282431)` line:
```python
py.test tests/foreman/api/test_docker.py -k test_create_container_content_view
============================= test session starts ==============================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/andrii/workspace/robottelo, inifile: 
collected 45 items

tests/foreman/api/test_docker.py .

 44 tests deselected by '-k DockerContainersTestCase and test_create_container_content_view' 
================== 1 passed, 44 deselected in 231.87 seconds ===================

py.test tests/foreman/ui/test_docker.py -k test_create_container_compute_resource
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/andrii/workspace/robottelo, inifile: 
collected 44 items 

tests/foreman/ui/test_docker.py .s

========= 42 tests deselected by '-ktest_create_container_compute_resource' ==========
================ 1 passed, 1 skipped, 42 deselected in 289.59 seconds ================

py.test tests/foreman/cli/test_docker.py -k test_create_container_content_view
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/andrii/workspace/robottelo, inifile: 
collected 51 items 

tests/foreman/cli/test_docker.py .

=========== 50 tests deselected by '-ktest_create_container_content_view' ============
===================== 1 passed, 50 deselected in 423.61 seconds ======================
```

So after BZ1282431 is fixed docker container tests should be fine.
Unfortunately, couldn't verify any content host test - keep failing on enabling RH repo. That's not related to this PR and is probably caused by some manifest issue, so i won't block this pr with `DONOTMERGE` tag.

oshtaier commented on robottelo/helpers.py:
```diff
@@ -143,3 +143,53 @@ def read_data_file(filename):
+        :return: None.
+        :raises AssertionError: If katello-ca wasn't removed.
+        """
+        ssh.command(
```
As per @sthirugn comment:
>Nope. Unregistering won't fix this. You need to clean up the red hat subscription management server things as shown in https://github.com/SatelliteQE/automation-tools/blob/master/automation_tools/__init__.py#L1284

Could you, please analyze that and provide summary of your investigation there?

abalakh replied:
Maybe I got something wrong, but later at that disccussion @omaciel and @sthirugn agreed that it's save and enough to simply delete `rpm` and that's it, didn't they?

As per their comments:
> It doesn't hurt us at all.

 and
> @omaciel is right. Since we are not getting sat tools in the sat host itself. But we do it from the satellite application which is configured with cdn info already. Registering the satellite host to itself should not be a problem for the sync tests.

Although I agree it's better to double check. Will check some random sync & longrun tests.

omaciel replied:
I'm 99.99% sure that it doesn't hurt to remove `katello-ca`. The 'side effect' is that if you wanted to yum install anything in the Satellite system itself, you'd get a big fat zero since the Satellite would be configured against itself, unless the Satellite has been registered against an org with proper content.

abalakh replied:
@omaciel just to confirm, but removing katello-ca you mean just removal of corresponding rpm or all the steps from [clean_rhsm()](https://github.com/SatelliteQE/automation-tools/blob/master/automation_tools/__init__.py#L1284) ?

omaciel replied:
Good question. To be thorough we'd have to fix rhsm.conf too but I'm wondering if that is really important... let's ask folks today during our meeting.

oshtaier commented:
Pending question

omaciel commented:
My only question is: I thought we were going to use `setUpModule` and `tearDownModule` to only make these calls once per module. Are we doing this at the class level because we don't have many tests that require this?

abalakh commented:
@omaciel exactly. Right now all our tests which require katello-ca are inside of `DockerContainersTestCase` class with no exclusions, so it's better to use `setUpClass`. If/when we add katello-ca tests dependent from katello-ca to another classes then i guess it'll be a good time to move to `setUpModule` not to duplicate the same code for numerous classes.

abalakh commented:
OK, figured out what's wrong with content_host tests. Fixed & rebased.
Just some random test result:
```python
py.test tests/foreman/cli/test_contenthost.py -k test_ch_package_install
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
rootdir: /home/andrii/workspace/robottelo, inifile: 
collected 22 items 

tests/foreman/cli/test_contenthost.py .

================= 21 tests deselected by '-ktest_ch_package_install' =================
===================== 1 passed, 21 deselected in 371.93 seconds ======================
```

elyezer commented on robottelo/vm.py:
```diff
@@ -212,27 +212,6 @@ def install_katello_agent(self):
         if result.return_code != 0:
             raise VirtualMachineError('Failed to install katello-agent')
 
-    def install_katello_cert(self):
```
What about leaving this method, but calling the helper under the hood?

elyezer replied:
Maybe we should add the remove equivalent too.

abalakh replied:
Makes sence for me :+1: 
I don't like how it looks right now, as for vm's we don't have a valid hostname and are forced to pass ip address to hostname argument which is ugly. Leaving a helper will resolve that.

elyezer commented:
Added some non-block suggestions.

Will let either @omaciel or @oshtaier take a look and merge if ok.

ACK

abalakh commented:
Updated & rebased.
As suggested, returned `vm.install_katello_cert()` (and renamed it to `vm.install_katello_ca()` for consistency) and added `vm.remove_katello_ca()` helpers.
That gives us at least 2 advantages:
* No need to pass `vm.ip_addr` to `hostname=` argument, which was ugly workaround for vms:
```python
# Before
install_katello_ca(hostname=client.ip_addr)
# Now
client.install_katello_ca()
```
* As `vm.install_katello_ca()` is inside `VirtualMachine` class, now we can raise commonly used for vms `VirtualMachineError` instead of `AssertionError` which isn't used anywhere else.

abalakh commented:
There's still open question regarding is removing of the `katello-ca.rpm` enough in `remove_katello_ca()` or some extra steps like fixing some configs/certificates are required.

sthirugn commented:
@abalakh If you want to keep this 100% clean, you need to do everything mentioned in automation-tools.

abalakh commented:
@sthirugn thanks for clarification, i've updated `remove_katello_ca` according to [clean_rhsm()](https://github.com/SatelliteQE/automation-tools/blob/master/automation_tools/__init__.py#L1296)
Some tests results:
```python
py.test -v tests/foreman/cli/test_contenthost.py -k test_ch_package_install
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
collected 22 items 

tests/foreman/cli/test_contenthost.py::TestCHKatelloAgent::test_ch_package_install <- robottelo/decorators.py PASSED

================= 21 tests deselected by '-ktest_ch_package_install' =================
===================== 1 passed, 21 deselected in 369.14 seconds ======================

py.test -v tests/foreman/cli/test_docker.py -k test_create_container_content_view
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.8.2, py-1.4.30, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
collected 51 items 

tests/foreman/cli/test_docker.py::DockerContainersTestCase::test_create_container_content_view <- robottelo/decorators.py PASSED

=========== 50 tests deselected by '-ktest_create_container_content_view' ============
===================== 1 passed, 50 deselected in 127.45 seconds ======================
```

abalakh commented:
Now PR is ready for a final review and can be merged if no new comments :)

elyezer commented:
ACK

oshtaier commented:
I don't know what I am talking about, but just question. Could we use method from automation tool directly? and do not duplicate code letter by letter?

abalakh commented:
We do not have automation tools as a dependency, it was dropped by #2644 . Introducing a new dependency just for 1 func imo isn't a gread idea.
Also, it is not absolutely the same, for example original function don't have validation of commands results (and this is important to have) and can't raise an exception in case of error.

sthirugn commented:
+1 to @abalakh, we dont need to introduce automation-tools as a dependency here.

oshtaier commented:
and it seems that tests are passed, so ACK

elyezer commented:
@abalakh @oshtaier  also fabric doesn't play well with py.test because the way how py.test deal with stdout and logging capturing. Since we are planning to use py.test for the tier thing then we should avoid automation tools.

elyezer commented:
This PR is open for a long time and already got two ACKs. Merging it.
